codec/json.go

```diff
@@ -412,15 +412,18 @@ func (e *jsonEncDriver) quoteStr(s string) {
 			continue
 		}
 		c, size := utf8.DecodeRuneInString(s[i:])
-		if c == utf8.RuneError {
-			if size == 1 {
-				if start < i {
-					w.writestr(s[start:i])
-				}
-				w.writestr(`\ufffd`)
-				i++
-				start = i
+		if c == utf8.RuneError && size == 1 {
+			// https://golang.org/pkg/unicode/utf8/#DecodeRuneInString
+			// c=utf8.RuneError, size=0: empty input (shouldn't happen here since i < len(s))
+			// c=utf8.RuneError, size=1: invalid encoding (handle below by skipping a byte and writing the replacement character \ufffd
+			// c=utf8.RuneError, size=2,4: shouldn't ever be returned
+			// c=utf8.RuneError, size=3: literal '\ufffd' in input, this is ok
+			if start < i {
+				w.writestr(s[start:i])
 			}
+			w.writestr(`\ufffd`)
+			i++
+			start = i
```

Comment:
The key fix is that before, the case of `c, size == utf8.RuneError, 3` wasn't handled and instead resulted in an infinite loop.

Reply:
The fix and test looks okay for the replacement character case and invalid unicode code points.

But what about other unicode characters? How should it work? It seems that only line separator and paragraph separator cases are handled later on and if this does not match, the character is skipped? (`i += uint(size)`).

There is also a `TestJsonInvalidUnicode` but I don't see a test that would check for valid unicode characters marshalled into JSON.

Am I missing something?

Reply:
Also https://github.com/ugorji/go/blob/a4365be44eb789f34688916df4ce68edc2541ae2/codec/codec_test.go#L3335 comment should probably be removed...

Reply:
Invalid unicode was already handled; that's the `c, size == utf8.RuneError, 1` case. What's added here is an actual `\ufffd` replacement character in the input, which is valid Unicode. And note that other unicode characters are handled similar to ASCII -- incrementing `i` without changing `start`. At the end of the loop, or before writing any output, this encoder flushes all the bytes from `start` to `i`.

I suspect you are right about that comment, I think it can go.